File: crates/taflow-metrics/src/metrics/average_win_test.rs
use super::average_win::AverageWin;
use crate::{MetricInputKind, NanPolicy};

fn assert_close(actual: f64, expected: f64) {
    assert!((actual - expected).abs() <= 1e-12, "{actual} != {expected}");
}

#[test]
fn computes_mean_of_strictly_positive_returns_and_preserves_lifecycle() {
    let values = [0.02, -0.01, 0.0, 0.03, -0.025, 0.01];
    let expected = (0.02 + 0.03 + 0.01) / 3.0;
    let mut state = AverageWin::new(NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();

    assert_eq!(state.value(), None);
    assert_eq!(state.append(values[0]).unwrap(), Some(values[0]));
    state.extend(&values[1..3]).unwrap();
    state.append(values[3]).unwrap();
    assert_close(state.extend(&values[4..]).unwrap().unwrap(), expected);
    assert_close(state.compute().unwrap(), expected);
    assert_eq!(state.len(), values.len());

    state.reset();
    assert!(state.is_empty());
    assert_eq!(state.value(), None);
    assert_close(state.extend(&values).unwrap().unwrap(), expected);
}

#[test]
fn raw_period_pnl_and_closed_trades_are_not_converted() {
    let observations = [100.0, -40.0, 0.0, 20.0, -10.0];
    let expected = 60.0;

    let mut pnl = AverageWin::new(NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_pnl(&[])?;
            Ok(state)
        })
        .unwrap();
    assert_close(pnl.extend(&observations).unwrap().unwrap(), expected);

    let mut trades = AverageWin::new(NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_trades(&[])?;
            Ok(state)
        })
        .unwrap();
    assert_close(trades.extend(&observations).unwrap().unwrap(), expected);
}

#[test]
fn losses_and_breakevens_do_not_create_a_result() {
    let mut state = AverageWin::new(NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    state.extend(&[-0.10, 0.0, -0.02]).unwrap();
    assert_eq!(state.len(), 3);
    assert_eq!(state.value(), None);

    assert_eq!(state.append(0.05).unwrap(), Some(0.05));
}

#[test]
fn missing_and_invalid_values_follow_the_input_contract() {
    let mut omit = AverageWin::new(NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    omit.extend(&[f64::NAN, 0.02, -0.01]).unwrap();
    assert_eq!(omit.len(), 2);
    assert_eq!(omit.value(), Some(0.02));

    let mut raise = AverageWin::new(NanPolicy::Raise)
        .and_then(|mut state| {
            state.from_returns(&[])?;
            Ok(state)
        })
        .unwrap();
    raise.append(0.01).unwrap();
    assert!(raise.append(f64::NAN).is_err());
    assert_eq!(raise.len(), 1);
    assert_eq!(raise.value(), Some(0.01));

    assert!(omit.append(f64::INFINITY).is_err());
    assert!(omit.append(-1.01).is_err());
    assert_eq!(omit.len(), 2);
}

#[test]
fn rejects_return_converters_outside_the_declared_domains() {
    assert!(AverageWin::new(NanPolicy::Omit)
        .and_then(|mut state| {
            state.append(0.0)?;
            Ok(state)
        })
        .is_err());
    assert!(AverageWin::new(NanPolicy::Omit)
        .and_then(|mut state| {
            state.append(0.0)?;
            Ok(state)
        })
        .is_err());
    assert!(AverageWin::new(NanPolicy::Omit)
        .and_then(|mut state| {
            state.append(0.0)?;
            Ok(state)
        })
        .is_err());
}

File: tests/metrics/average_win_test.py
from __future__ import annotations
import numpy as np
import pandas as pd
import pytest
import quantstats.stats as qs
from taflow.metrics.average_win import AverageWin

def _quantstats_average_win(values: np.ndarray) -> float | None:
    expected = float(qs.avg_win(pd.Series(values), aggregate=None, compounded=True, prepare_returns=False))
    return expected if np.isfinite(expected) else None

@pytest.mark.parametrize('returns', [np.array([0.02, -0.01, 0.0, 0.03, -0.025, 0.01]), np.array([0.1]), np.zeros(32), np.linspace(-0.04, 0.05, 101), np.array([0.15, -0.1, -0.1, 0.25, -0.3, 0.1]), np.array([np.nan, 0.02, -0.03, np.nan, 0.01]), np.random.default_rng(20260811).normal(0.0004, 0.012, 513)])
def test_average_win_matches_quantstats(returns: np.ndarray) -> None:
    actual = AverageWin().from_returns(returns).compute()
    expected = _quantstats_average_win(returns)
    if expected is None:
        assert actual is None
    else:
        assert actual == pytest.approx(expected, rel=1e-12, abs=1e-14)

def test_average_win_raw_pnl_and_trade_input_methods_match_quantstats() -> None:
    observations = np.array([100.0, -40.0, 0.0, 20.0, -10.0])
    expected = _quantstats_average_win(observations)
    assert expected == 60.0
    assert AverageWin().from_pnl(observations).compute() == expected
    assert AverageWin().from_trades(observations).compute() == expected

def test_average_win_lifecycle_is_chunk_and_reset_invariant() -> None:
    values = np.array([0.02, -0.01, 0.0, 0.03, -0.025, 0.01])
    expected = AverageWin().from_returns(values).compute()
    state = AverageWin().from_returns([])
    assert state.value is None
    assert state.append(values[0]) is state
    assert state.value == values[0]
    assert state.extend(values[1:3]) is state
    assert state.extend(values[3:]) is state
    assert state.compute() == pytest.approx(expected)
    assert len(state) == len(values)
    assert state.reset() is state
    assert len(state) == 0
    assert state.value is None
    assert state.extend(values).compute() == pytest.approx(expected)

def test_average_win_boundaries_missing_and_invalid_values() -> None:
    assert AverageWin().from_returns([]).compute() is None
    assert AverageWin().from_returns([-0.25, 0.0]).compute() is None
    assert AverageWin().from_returns([0.25]).compute() == 0.25
    assert len(AverageWin().from_returns([np.nan, 0.01, -0.02])) == 2
    with pytest.raises(ValueError):
        AverageWin(nan_policy='raise').from_returns([np.nan])
    with pytest.raises(ValueError):
        AverageWin().from_returns([np.inf])
    with pytest.raises(ValueError):
        AverageWin().from_returns([-1.01])
    assert AverageWin().from_pnl([-1000.0, 50.0]).compute() == 50.0
    assert AverageWin().from_trades([-1000.0, 50.0]).compute() == 50.0

def test_average_win_requires_semantic_input_method() -> None:
    metric = AverageWin()
    with pytest.raises(ValueError):
        metric.append(0.01)
